include/quic/gnutls_crypto.hpp

```diff
@@ -25,36 +102,78 @@ namespace oxen::quic
             unsigned int incoming,
             const gnutls_datum_t* msg)>;
 
+    inline constexpr size_t CERT_HEADER_SIZE = 12;
     inline constexpr size_t GNUTLS_KEY_SIZE = 32;  // for now, only supporting Ed25519 keys (32 bytes)
     inline constexpr size_t GNUTLS_SECRET_KEY_SIZE = 64;
 
-    using gnutls_key = std::array<unsigned char, GNUTLS_KEY_SIZE>;
+    // These bytes mean "this is a raw Ed25519 private key" in ASN.1 (or something like that)
+    inline const std::string ASN_ED25519_SEED_PREFIX = oxenc::from_hex("302e020100300506032b657004220420"sv);
+    // These bytes mean "this is a raw Ed25519 public key" in ASN.1 (or something like that)
+    inline const std::string ASN_ED25519_PUBKEY_PREFIX = oxenc::from_hex("302a300506032b6570032100"sv);
+
+    struct gnutls_key
+    {
+      private:
+        std::array<unsigned char, GNUTLS_KEY_SIZE> buf{};
+
+        gnutls_key(const unsigned char* data, size_t size) { write(data, size); }
 
-    // arguments: remote pubkey, ALPN
-    using gnutls_key_verify_callback = std::function<bool(const gnutls_key&, const std::string_view& alpn)>;
+      public:
+        gnutls_key() = default;
+        gnutls_key(std::string_view data) : gnutls_key{convert_sv<unsigned char>(data)} {}
+        gnutls_key(ustring_view data) : gnutls_key{data.data(), data.size()} {}
+
+        //  Writes to the internal buffer holding the gnutls key
+        void write(const unsigned char* data, size_t size)
+        {
+            if (size != GNUTLS_KEY_SIZE)
+                throw std::invalid_argument{"GNUTLS key must be 32 bytes!"};
+
+            std::memcpy(buf.data(), data, size);
+        }
+
+        ustring_view view() const { return {buf.data(), buf.size()}; }
+
+        gnutls_key(const gnutls_key& other) { *this = other; }
+
+        gnutls_key& operator=(const gnutls_key& other)
+        {
+            buf = other.buf;
+            return *this;
+        }
+
+        void operator()(ustring_view data) { write(data.data(), data.size()); }
+
+        explicit operator bool() const { return not buf.empty(); }
+
+        bool operator==(const gnutls_key& other) const { return buf == other.buf; }
```

Comment:
It's pretty common to also add a `!=` operator as a matter of practice when you have a `==` operator, which is just defined by the `==` operator, i.e. just does `return !(*this == other);`

(C++20 adds the spaceship (`<=>`) operator which returns a memcmp-like -1/0/1 value, so you can define just that *one* and automatically get all of `==`, `!=`, `<`, `<=`, `>`, and `>=` provided automatically for you; in fact I think much of the STL got redefined to use that when in C++20 mode).

Reply:
Nice thanks added